src/k_smallest.rs

```diff
@@ -1,20 +1,245 @@
-use alloc::collections::BinaryHeap;
-use core::cmp::Ord;
+use core::cmp::{Ord, Ordering, Reverse};
+use core::mem::{replace, transmute, MaybeUninit};
 
-pub(crate) fn k_smallest<T: Ord, I: Iterator<Item = T>>(mut iter: I, k: usize) -> BinaryHeap<T> {
-    if k == 0 { return BinaryHeap::new(); }
+fn k_smallest_dynamic<T, I: Iterator<Item = T>>(
+    iter: I,
+    k: usize,
+    order: impl Fn(&T, &T) -> Ordering,
+) -> std::vec::IntoIter<T> {
```

Comment:
Using `std::` instead of `alloc::` is going to break users who can't pull in `std` but can pull in `alloc`.

Reply:
Good point, been fixed to use `alloc`